### PR TITLE
Make support for quadruple precision optional

### LINKED ISSUE #485: API Specification for stdlib_kinds
Documentation for the kinds available from `stdlib_kinds` is currently missing.

### LINKED ISSUE #527: stdlib build error with fpm (macOS Apple Silicon)
### Description

Hi

I have attempted to build and install stdlib using `fpm build --profile release` but get the following error message:
```console
 + mkdir -p build/dependencies
 + mkdir -p build/gfortran_5F75F7C92365B9B9/stdlib
 + gfortran -c ././src/stdlib_kinds.f90 -O3 -Wimplicit-interface -fPIC -fmax-errors=1 -funroll-loops -fcoarray=single -J build/gfortran_5F75F7C92365B9B9/stdlib -I build/gfortran_5F75F7C92365B9B9/stdlib  -o build/gfortran_5F75F7C92365B9B9/stdlib/src_stdlib_kinds.f90.o
 + gfortran -c ././src/stdlib_system.F90 -O3 -Wimplicit-interface -fPIC -fmax-errors=1 -funroll-loops -fcoarray=single -J build/gfortran_5F75F7C92365B9B9/stdlib -I build/gfortran_5F75F7C92365B9B9/stdlib  -o build/gfortran_5F75F7C92365B9B9/stdlib/src_stdlib_system.F90.o
 + gfortran -c test/test_sleep.f90 -O3 -Wimplicit-interface -fPIC -fmax-errors=1 -funroll-loops -fcoarray=single -J build/gfortran_5F75F7C92365B9B9/stdlib -I build/gfortran_5F75F7C92365B9B9/stdlib  -o build/gfortran_5F75F7C92365B9B9/stdlib/test_test_sleep.f90.o
 + gfortran -c ././src/stdlib_ascii.f90 -O3 -Wimplicit-interface -fPIC -fmax-errors=1 -funroll-loops -fcoarray=single -J build/gfortran_5F75F7C92365B9B9/stdlib -I build/gfortran_5F75F7C92365B9B9/stdlib  -o build/gfortran_5F75F7C92365B9B9/stdlib/src_stdlib_ascii.f90.o
 + gfortran -c ././src/stdlib_bitsets.f90 -O3 -Wimplicit-interface -fPIC -fmax-errors=1 -funroll-loops -fcoarray=single -J build/gfortran_5F75F7C92365B9B9/stdlib -I build/gfortran_5F75F7C92365B9B9/stdlib  -o build/gfortran_5F75F7C92365B9B9/stdlib/src_stdlib_bitsets.f90.o
 + gfortran -c ././src/stdlib_optval.f90 -O3 -Wimplicit-interface -fPIC -fmax-errors=1 -funroll-loops -fcoarray=single -J build/gfortran_5F75F7C92365B9B9/stdlib -I build/gfortran_5F75F7C92365B9B9/stdlib  -o build/gfortran_5F75F7C92365B9B9/stdlib/src_stdlib_optval.f90.o
././src/stdlib_optval.f90:70:12:

   70 |     real(qp), intent(in), optional :: x
      |            1
Error: Kind -1 not supported for type REAL at (1)
compilation terminated due to -fmax-errors=1.
 <ERROR> Compilation failed for object "src_stdlib_optval.f90.o"
STOP 1
```

Versions being used:
```console
% fpm -v
Version:     0.3.0, alpha
Program:     fpm(1)
Description: A Fortran package manager and build system
Home Page:   https://github.com/fortran-lang/fpm
License:     MIT
OS Type:     macOS

% gfortran --version
GNU Fortran (Homebrew GCC 11.2.0) 11.1.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

% sw_vers
ProductName:	macOS
ProductVersion:	11.6
BuildVersion:	20G165
```

Steps used to build:
```
git clone https://github.com/fortran-lang/stdlib
cd stdlib
git checkout stdlib-fpm
fpm build --profile release
```

I am running on a new Apple computer using Apple Silicon M1 CPU - which maybe the cause? I have run into other build errors previously to do with `REAL` issues - so it maybe the same problem?

For reference in case it helps: https://github.com/urbanjost/M_strings/issues/2

Thanks for any help / advice!

Simon

### Expected Behaviour

That stdlib builds OK using `fpm` using `gfortran` on macOS arm64.

### Version of stdlib

Current version obtained using the steps document above. 

### Platform and Architecture

macOS (arm64) Apple Silicon M1 Big Sur version 11.6

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by awvwgk**
- use checks in CMake to automatically determine whether quadruple precision is available
- default to values in kind we run from the manual makefile don't want to run checks
- skip all tests explicitly referencing the quadruple precision kind parameter
- add check for extended double precision (available with GCC)

Closes #527 
Closes #485
Closes #567

### CODE REVIEW

**Comment by awvwgk** [comment]
I think the manual makefile is accidentally picking up the fypp preprocessed source files generated the in-tree CMake build. Maybe we should split the workflow for the different builds?

**Comment by awvwgk** [comment]
Sorry for the mess with the Makefiles. I wish there would be a way to make this cleaner, but it should be working again now. Bad news is we are going to loose a couple of tests due to https://github.com/fortran-lang/stdlib/issues/534.

File: doc/specs/stdlib_kinds.md [hunk 1]
```diff
@@ -16,17 +16,24 @@ The `stdlib_kinds` module provides kind parameters for the Fortran intrinsic dat
 
 ### `sp`
 
-Alias for intrinsic named constant `real32` imported from `iso_fortran_env`.
+Provides real kind parameter for floating point numbers with a minimal precision of 6 significant digits.
```
Comment (certik) [inline doc/specs/stdlib_kinds.md:19]: Should we document what the constants `sp` stands for here? Such as

`sp` stands for "single precision". Is `xdp` "extended double precision"?

**Review by certik — APPROVED** [review]
I think this looks good. Left a minor comment that can be fixed by another PR.

I like to think of our CMake build system as temporary, we will eventually generate it from `fpm`. What changes to `fpm` would be needed, so that it could (in the future) generate the CMake files in this PR?

**Comment by awvwgk** [comment]
> What changes to `fpm` would be needed, so that it could (in the future) generate the CMake files in this PR?

Creating CMake files from fpm is actually not that hard, at some point I had put together a version of fpm which could write meson and CMake build files (the branch might still be in my fork). The main issue I encountered at this point was that handling dependencies is not simple, because it was not clear how nested subprojects could be handled reliably in CMake (duplicates in the dependency tree, cyclic dependencies), therefore I never really got the CMake build working for more than a standalone project.

Today, I would try it using generated find-modules and checking for the existence of imported targets, like we are currently doing in [stdlib-cmake-example](https://github.com/fortran-lang/stdlib-cmake-example).

Doing this for stdlib additionally requires preprocessor support (https://github.com/fortran-lang/fpm/issues/308, https://github.com/fortran-lang/fpm/issues/78) and potentially optional features (https://github.com/fortran-lang/fpm/issues/609).

**Comment by certik** [comment]
Looks good to me to merge.

**Review by milancurcic — APPROVED** [review]
Thank you!

**Comment by awvwgk** [comment]
Just saw I broke the fpm deployment and the manual makefile build on all other branches. I opened a hotfix for this in #567 or we can merge this PR to fix everything up again.

Sorry for the mess.

**Comment by milancurcic** [comment]
Yes, if this fixes everything, we can just merge this one and close #567.

**Comment by awvwgk** [comment]
Let's add a pin for test-drive here as well, while this requires manual updates from time to time, it will help to mitigate similar issues in the future. I'll merge this PR once the CI passes.

**Pull request merged**